File: lib/ruby2js/rails/migration_sql.rb
# frozen_string_literal: true

# Generates SQL from Rails migration files for D1 and other SQL databases
# Uses Ruby2JS's parser to parse the migration Ruby code

require 'ruby2js'
require 'ruby2js/inflector'

module Ruby2JS
  module Rails
    class MigrationSQL
      # Generate SQL for all migrations in a directory
      # Returns a hash with :sql (combined SQL) and :migrations (array of parsed migrations)
      def self.generate_all(migrate_dir)
        return { sql: '', migrations: [] } unless File.exist?(migrate_dir)

        migrations = []
        sql_parts = []

        sql_parts << "-- Ruby2JS Generated Migrations"
        sql_parts << "-- Generated at: #{Time.now.utc}"
        sql_parts << ""
        sql_parts << "-- Schema migrations tracking table"
        sql_parts << "CREATE TABLE IF NOT EXISTS schema_migrations (version TEXT PRIMARY KEY);"
        sql_parts << ""

        Dir.glob(File.join(migrate_dir, '*.rb')).sort.each do |path|
          basename = File.basename(path, '.rb')
          version = basename.split('_').first

          migration = self.parse_migration(path)
          next unless migration

          sql = self.generate_sql(migration[:statements], version)
          sql_parts << sql if sql && !sql.empty?

          migrations << { version: version, filename: basename }
        end

        { sql: sql_parts.join("\n"), migrations: migrations }
      end

      # Parse a single migration file and extract statements
      def self.parse_migration(path)
        source = File.read(path)
        ast, _ = Ruby2JS.parse(source)
        return nil unless ast

        statements = []
        self.extract_statements(ast, statements)

        { path: path, statements: statements }
      end

      # Generate SQL from migration statements
      def self.generate_sql(statements, version)
        return '' if statements.empty?

        sql_lines = []
        sql_lines << "-- Migration: #{version}"

        statements.each do |stmt|
          case stmt[:type]
          when :create_table
            sql_lines << self.create_table_sql(stmt)
          when :add_index
            sql_lines << self.add_index_sql(stmt)
          when :add_column
            sql_lines << self.add_column_sql(stmt)
          when :remove_column
            sql_lines << self.remove_column_sql(stmt)
          when :drop_table
            sql_lines << self.drop_table_sql(stmt)
          end
        end

        # Add version tracking
        sql_lines << "INSERT INTO schema_migrations (version) VALUES ('#{version}') ON CONFLICT DO NOTHING;"
        sql_lines << ""

        sql_lines.join("\n")
      end

      def self.extract_statements(node, statements)
        return unless node

        case node.type
        when :class
          # Check if this is a migration class
          name, parent, body = node.children
          if self.migration_class?(parent)
            self.extract_from_body(body, statements)
          end
        when :begin
          node.children.each { |child| self.extract_statements(child, statements) }
        end
      end

      def self.migration_class?(node)
        return false unless node

        if node.type == :send && node.children[1] == :[]
          const = node.children[0]
          return self.migration_const?(const)
        elsif node.type == :const
          return self.migration_const?(node)
        end

        false
      end

      def self.migration_const?(node)
        return false unless node&.type == :const
        children = node.children
        return false unless children.length == 2

        parent = children[0]
        name = children[1]

        parent&.type == :const &&
          parent.children[0].nil? &&
          parent.children[1] == :ActiveRecord &&
          name == :Migration
      end

      def self.extract_from_body(body, statements)
        return unless body

        children = body.type == :begin ? body.children : [body]

        children.each do |child|
          next unless child

          if child.type == :def
            method_name = child.children[0]
            if method_name == :change || method_name == :up
              self.extract_from_method(child.children[2], statements)
            end
          end
        end
      end

      def self.extract_from_method(body, statements)
        return unless body

        children = body.type == :begin ? body.children : [body]

        children.each do |child|
          next unless child

          case child.type
          when :block
            self.extract_block(child, statements)
          when :send
            self.extract_send(child, statements)
          end
        end
      end

      def self.extract_block(node, statements)
        call, block_args, body = node.children
        return unless call.type == :send

        target, method, *args = call.children
        return unless target.nil?

        case method
        when :create_table
          self.extract_create_table(args, block_args, body, statements)
        end
      end

      def self.extract_send(node, statements)
        target, method, *args = node.children
        return unless target.nil?

        case method
        when :add_index
          self.extract_add_index(args, statements)
        when :add_column
          self.extract_add_column(args, statements)
        when :remove_column
          self.extract_remove_column(args, statements)
        when :drop_table
          self.extract_drop_table(args, statements)
        end
      end

      def self.extract_create_table(args, block_args, body, statements)
        return if args.empty?

        table_name = self.extract_string_value(args[0])
        return unless table_name

        options = self.extract_table_options(args)
        columns = []
        foreign_keys = []

        # Add primary key unless id: false
        unless options[:id] == false
          columns << {
            name: 'id',
            type: 'integer',
            primaryKey: true,
            autoIncrement: true
          }
        end

        # Process column definitions
        if body
          column_children = body.type == :begin ? body.children : [body]

          column_children.each do |child|
            next unless child&.type == :send

            result = self.extract_column(child, table_name)
            if result
              columns << result[:column] if result[:column]
              columns.push(*result[:columns]) if result[:columns]
              foreign_keys << result[:foreign_key] if result[:foreign_key]
            end
          end
        end

        statements << {
          type: :create_table,
          table: table_name,
          columns: columns,
          foreign_keys: foreign_keys
        }
      end

      def self.extract_column(node, table_name)
        target, method, *args = node.children

        return nil unless target&.type == :lvar && target.children[0] == :t

        case method
        when :timestamps
          {
            columns: [
              { name: 'created_at', type: 'datetime', null: false },
              { name: 'updated_at', type: 'datetime', null: false }
            ]
          }
        when :references, :belongs_to
          self.extract_references(args, table_name)
        else
          self.extract_regular_column(method, args)
        end
      end

      def self.extract_regular_column(type, args)
        return nil if args.empty?

        column_name = self.extract_string_value(args[0])
        return nil unless column_name

        column = {
          name: column_name,
          type: type.to_s
        }

        args[1..-1].each do |arg|
          next unless arg.type == :hash

          arg.children.each do |pair|
            key = pair.children[0]
            value = pair.children[1]
            next unless key.type == :sym

            case key.children[0]
            when :null
              column[:null] = (value.type != :false)
            when :default
              column[:default] = self.extract_default_value(value)
            when :limit
              column[:limit] = value.children[0] if value.type == :int
            end
          end
        end

        { column: column }
      end

      def self.extract_references(args, table_name)
        return nil if args.empty?

        ref_name = self.extract_string_value(args[0])
        return nil unless ref_name

        column_name = "#{ref_name}_id"
        column = {
          name: column_name,
          type: 'integer',
          null: false
        }
        foreign_key = nil

        args[1..-1].each do |arg|
          next unless arg.type == :hash

          arg.children.each do |pair|
            key = pair.children[0]
            value = pair.children[1]
            next unless key.type == :sym

            case key.children[0]
            when :null
              column[:null] = true if value.type == :true
            when :foreign_key
              if value.type == :true
                ref_table = Ruby2JS::Inflector.pluralize(ref_name)
                foreign_key = {
                  column: column_name,
                  references_table: ref_table,
                  references_column: 'id'
                }
              end
            end
          end
        end

        { column: column, foreign_key: foreign_key }
      end

      def self.extract_add_index(args, statements)
        return if args.length < 2

        table_name = self.extract_string_value(args[0])
        return unless table_name

        columns = if args[1].type == :array
                    args[1].children.map { |c| self.extract_string_value(c) }.compact
                  else
                    [self.extract_string_value(args[1])].compact
                  end

        return if columns.empty?

        options = {}
        args[2..-1].each do |arg|
          next unless arg&.type == :hash

          arg.children.each do |pair|
            key = pair.children[0]
            value = pair.children[1]
            next unless key.type == :sym

            case key.children[0]
            when :name
              options[:name] = self.extract_string_value(value)
            when :unique
              options[:unique] = value.type == :true
            end
          end
        end

        statements << {
          type: :add_index,
          table: table_name,
          columns: columns,
          options: options
        }
      end

      def self.extract_add_column(args, statements)
        return if args.length < 3

        table_name = self.extract_string_value(args[0])
        column_name = self.extract_string_value(args[1])
        column_type = self.extract_string_value(args[2])

        return unless table_name && column_name && column_type

        statements << {
          type: :add_column,
          table: table_name,
          column: column_name,
          column_type: column_type
        }
      end

      def self.extract_remove_column(args, statements)
        return if args.length < 2

        table_name = self.extract_string_value(args[0])
        column_name = self.extract_string_value(args[1])

        return unless table_name && column_name

        statements << {
          type: :remove_column,
          table: table_name,
          column: column_name
        }
      end

      def self.extract_drop_table(args, statements)
        return if args.empty?

        table_name = self.extract_string_value(args[0])
        return unless table_name

        statements << {
          type: :drop_table,
          table: table_name
        }
      end

      def self.extract_table_options(args)
        options = {}

        args[1..-1].each do |arg|
          next unless arg&.type == :hash

          arg.children.each do |pair|
            key = pair.children[0]
            value = pair.children[1]
            next unless key.type == :sym

            case key.children[0]
            when :id
              options[:id] = value.type != :false
            end
          end
        end

        options
      end

      def self.extract_string_value(node)
        case node&.type
        when :str then node.children[0]
        when :sym then node.children[0].to_s
        else nil
        end
      end

      def self.extract_default_value(node)
        case node.type
        when :str then node.children[0]
        when :int, :float then node.children[0]
        when :true then true
        when :false then false
        when :nil then nil
        else nil
        end
      end

      # SQL generation methods

      def self.create_table_sql(stmt)
        columns_sql = stmt[:columns].map do |col|
          self.column_def_sql(col)
        end

        # Add foreign key constraints
        stmt[:foreign_keys].compact.each do |fk|
          columns_sql << "FOREIGN KEY (#{fk[:column]}) REFERENCES #{fk[:references_table]}(#{fk[:references_column]})"
        end

        "CREATE TABLE IF NOT EXISTS #{stmt[:table]} (\n  #{columns_sql.join(",\n  ")}\n);"
      end

      def self.column_def_sql(col)
        parts = [col[:name]]
        parts << self.sql_type(col[:type])

        if col[:primaryKey]
          parts << "PRIMARY KEY"
          parts << "AUTOINCREMENT" if col[:autoIncrement]
        end

        if col.key?(:null) && col[:null] == false
          parts << "NOT NULL"
        end

        if col.key?(:default)
          parts << "DEFAULT #{self.sql_value(col[:default])}"
        end

        parts.join(' ')
      end

      def self.sql_type(type)
        case type.to_s
        when 'string' then 'TEXT'
        when 'text' then 'TEXT'
        when 'integer' then 'INTEGER'
        when 'float', 'decimal' then 'REAL'
        when 'boolean' then 'INTEGER'
        when 'datetime', 'timestamp' then 'TEXT'
        when 'date' then 'TEXT'
        when 'time' then 'TEXT'
        when 'binary', 'blob' then 'BLOB'
        else 'TEXT'
        end
      end

      def self.sql_value(value)
        case value
        when String then "'#{value.gsub("'", "''")}'"
        when Integer, Float then value.to_s
        when true then '1'
        when false then '0'
        when nil then 'NULL'
        else "'#{value}'"
        end
      end

      def self.add_index_sql(stmt)
        unique = stmt[:options][:unique] ? 'UNIQUE ' : ''
        name = stmt[:options][:name] || "index_#{stmt[:table]}_on_#{stmt[:columns].join('_')}"
        columns = stmt[:columns].join(', ')

        "CREATE #{unique}INDEX IF NOT EXISTS #{name} ON #{stmt[:table]} (#{columns});"
      end

      def self.add_column_sql(stmt)
        "ALTER TABLE #{stmt[:table]} ADD COLUMN #{stmt[:column]} #{self.sql_type(stmt[:column_type])};"
      end

      def self.remove_column_sql(stmt)
        # SQLite doesn't support DROP COLUMN before 3.35.0
        # D1 uses SQLite, so we need to handle this carefully
        "-- Note: SQLite may not support DROP COLUMN\n-- ALTER TABLE #{stmt[:table]} DROP COLUMN #{stmt[:column]};"
      end

      def self.drop_table_sql(stmt)
        "DROP TABLE IF EXISTS #{stmt[:table]};"
      end
    end
  end
end

File: packages/ruby2js-rails/lib/migration_sql.js
import { Ruby2JS } from "ruby2js/ruby2js.js";

export class MigrationSQL {
  // Generate SQL for all migrations in a directory
  // Returns a hash with :sql (combined SQL) and :migrations (array of parsed migrations)
  static generate_all(migrate_dir) {
    if (!File.exist(migrate_dir)) return {sql: "", migrations: []};
    let migrations = [];
    let sql_parts = [];
    sql_parts.push("-- Ruby2JS Generated Migrations");
    sql_parts.push(`-- Generated at: ${Time.now.utc}`);
    sql_parts.push("");
    sql_parts.push("-- Schema migrations tracking table");
    sql_parts.push("CREATE TABLE IF NOT EXISTS schema_migrations (version TEXT PRIMARY KEY);");
    sql_parts.push("");

    for (let path of Dir.glob(File.join(migrate_dir, "*.rb")).sort()) {
      let basename = File.basename(path, ".rb");
      let version = basename.split("_")[0];
      let migration = this.parse_migration(path);
      if (!migration) continue;
      let sql = this.generate_sql(migration.statements, version);
      if (sql && sql.length != 0) sql_parts.push(sql);
      migrations.push({version, filename: basename})
    };

    return {sql: sql_parts.join("\n"), migrations}
  };

  // Parse a single migration file and extract statements
  static parse_migration(path) {
    let source = File.read(path);
    let [ast, _] = Ruby2JS.parse(source);
    if (!ast) return null;
    let statements = [];
    this.extract_statements(ast, statements);
    return {path, statements}
  };

  // Generate SQL from migration statements
  static generate_sql(statements, version) {
    if (statements.length == 0) return "";
    let sql_lines = [];
    sql_lines.push(`-- Migration: ${version}`);

    for (let stmt of statements) {
      switch (stmt.type) {
      case "create_table":
        sql_lines.push(this.create_table_sql(stmt));
        break;

      case "add_index":
        sql_lines.push(this.add_index_sql(stmt));
        break;

      case "add_column":
        sql_lines.push(this.add_column_sql(stmt));
        break;

      case "remove_column":
        sql_lines.push(this.remove_column_sql(stmt));
        break;

      case "drop_table":
        sql_lines.push(this.drop_table_sql(stmt))
      }
    };

    // Add version tracking
    sql_lines.push(`INSERT INTO schema_migrations (version) VALUES ('${version}') ON CONFLICT DO NOTHING;`);
    sql_lines.push("");
    return sql_lines.join("\n")
  };

  static extract_statements(node, statements) {
    let name, parent, body;
    if (!node) return;

    switch (node.type) {
    case "class":

      // Check if this is a migration class
      let [name, parent, body] = node.children;

      if (this.migration_class(parent)) {
        return this.extract_from_body(body, statements)
      };

      break;

    case "begin":

      for (let child of node.children) {
        this.extract_statements(child, statements)
      }
    }
  };

  static migration_class(node) {
    if (!node) return false;

    if (node.type == "send" && node.children[1] == "[]") {
      let $const = node.children[0];
      return this.migration_const($const)
    } else if (node.type == "const") {
      return this.migration_const(node)
    };

    return false
  };

  static migration_const(node) {
    if (node?.type != "const") return false;
    let children = node.children;
    if (children.length != 2) return false;
    let parent = children[0];
    let name = children[1];
    return parent?.type == "const" && parent.children[0] == null && parent.children[1] == "ActiveRecord" && name == "Migration"
  };

  static extract_from_body(body, statements) {
    if (!body) return;
    let children = body.type == "begin" ? body.children : [body];

    for (let child of children) {
      if (!child) continue;

      if (child.type == "def") {
        let method_name = child.children[0];

        if (method_name == "change" || method_name == "up") {
          this.extract_from_method(child.children[2], statements)
        }
      }
    }
  };

  static extract_from_method(body, statements) {
    if (!body) return;
    let children = body.type == "begin" ? body.children : [body];

    for (let child of children) {
      if (!child) continue;

      switch (child.type) {
      case "block":
        this.extract_block(child, statements);
        break;

      case "send":
        this.extract_send(child, statements)
      }
    }
  };

  static extract_block(node, statements) {
    let [call, block_args, body] = node.children;
    if (call.type != "send") return;
    let [target, method, ...args] = call.children;
    if (target != null) return;

    switch (method) {
    case "create_table":
      return this.extract_create_table(args, block_args, body, statements)
    }
  };

  static extract_send(node, statements) {
    let [target, method, ...args] = node.children;
    if (target != null) return;

    switch (method) {
    case "add_index":
      return this.extract_add_index(args, statements);

    case "add_column":
      return this.extract_add_column(args, statements);

    case "remove_column":
      return this.extract_remove_column(args, statements);

    case "drop_table":
      return this.extract_drop_table(args, statements)
    }
  };

  static extract_create_table(args, block_args, body, statements) {
    if (args.length == 0) return;
    let table_name = this.extract_string_value(args[0]);
    if (!table_name) return;
    let options = this.extract_table_options(args);
    let columns = [];
    let foreign_keys = [];

    // Add primary key unless id: false
    if (options.id != false) {
      columns.push({
        name: "id",
        type: "integer",
        primaryKey: true,
        autoIncrement: true
      })
    };

    // Process column definitions
    if (body) {
      let column_children = body.type == "begin" ? body.children : [body];

      for (let child of column_children) {
        if (child?.type != "send") continue;
        let result = this.extract_column(child, table_name);

        if (result) {
          if (result.column) columns.push(result.column);
          if (result.columns) columns.push(...result.columns);
          if (result.foreign_key) foreign_keys.push(result.foreign_key)
        }
      }
    };

    statements.push({
      type: "create_table",
      table: table_name,
      columns,
      foreign_keys
    })
  };

  static extract_column(node, table_name) {
    let [target, method, ...args] = node.children;
    if (target?.type != "lvar" || target.children[0] != "t") return null;

    switch (method) {
    case "timestamps":

      return {columns: [
        {name: "created_at", type: "datetime", null: false},
        {name: "updated_at", type: "datetime", null: false}
      ]};

    case "references":
    case "belongs_to":
      return this.extract_references(args, table_name);

    default:
      return this.extract_regular_column(method, args)
    }
  };

  static extract_regular_column(type, args) {
    if (args.length == 0) return null;
    let column_name = this.extract_string_value(args[0]);
    if (!column_name) return null;
    let column = {name: column_name, type: type.toString()};

    for (let arg of args.slice(1)) {
      if (arg.type != "hash") continue;

      for (let pair of arg.children) {
        let key = pair.children[0];
        let value = pair.children[1];
        if (key.type != "sym") continue;

        switch (key.children[0]) {
        case "null":
          column.null = value.type != "false";
          break;

        case "default":
          column.default = this.extract_default_value(value);
          break;

        case "limit":
          if (value.type == "int") column.limit = value.children[0]
        }
      }
    };

    return {column}
  };

  static extract_references(args, table_name) {
    if (args.length == 0) return null;
    let ref_name = this.extract_string_value(args[0]);
    if (!ref_name) return null;
    let column_name = `${ref_name}_id`;
    let column = {name: column_name, type: "integer", null: false};
    let foreign_key = null;

    for (let arg of args.slice(1)) {
      if (arg.type != "hash") continue;

      for (let pair of arg.children) {
        let key = pair.children[0];
        let value = pair.children[1];
        if (key.type != "sym") continue;

        switch (key.children[0]) {
        case "null":
          if (value.type == "true") column.null = true;
          break;

        case "foreign_key":

          if (value.type == "true") {
            let ref_table = Ruby2JS.Inflector.pluralize(ref_name);

            foreign_key = {
              column: column_name,
              references_table: ref_table,
              references_column: "id"
            }
          }
        }
      }
    };

    return {column, foreign_key}
  };

  static extract_add_index(args, statements) {
    if (args.length < 2) return;
    let table_name = this.extract_string_value(args[0]);
    if (!table_name) return;

    let columns = args[1].type == "array" ? args[1].children.map(c => (
      this.extract_string_value(c)
    )).filter(x => x != null) : [this.extract_string_value(args[1])].filter(x => (
      x != null
    ));

    if (columns.length == 0) return;
    let options = {};

    for (let arg of args.slice(2)) {
      if (arg?.type != "hash") continue;

      for (let pair of arg.children) {
        let key = pair.children[0];
        let value = pair.children[1];
        if (key.type != "sym") continue;

        switch (key.children[0]) {
        case "name":
          options.name = this.extract_string_value(value);
          break;

        case "unique":
          options.unique = value.type == "true"
        }
      }
    };

    statements.push({
      type: "add_index",
      table: table_name,
      columns,
      options
    })
  };

  static extract_add_column(args, statements) {
    if (args.length < 3) return;
    let table_name = this.extract_string_value(args[0]);
    let column_name = this.extract_string_value(args[1]);
    let column_type = this.extract_string_value(args[2]);
    if (!table_name || !column_name || !column_type) return;

    statements.push({
      type: "add_column",
      table: table_name,
      column: column_name,
      column_type
    })
  };

  static extract_remove_column(args, statements) {
    if (args.length < 2) return;
    let table_name = this.extract_string_value(args[0]);
    let column_name = this.extract_string_value(args[1]);
    if (!table_name || !column_name) return;

    statements.push({
      type: "remove_column",
      table: table_name,
      column: column_name
    })
  };

  static extract_drop_table(args, statements) {
    if (args.length == 0) return;
    let table_name = this.extract_string_value(args[0]);
    if (!table_name) return;
    statements.push({type: "drop_table", table: table_name})
  };

  static extract_table_options(args) {
    let options = {};

    for (let arg of args.slice(1)) {
      if (arg?.type != "hash") continue;

      for (let pair of arg.children) {
        let key = pair.children[0];
        let value = pair.children[1];
        if (key.type != "sym") continue;

        switch (key.children[0]) {
        case "id":
          options.id = value.type != "false"
        }
      }
    };

    return options
  };

  static extract_string_value(node) {
    switch (node?.type) {
    case "str":
      return node.children[0];

    case "sym":
      return node.children[0].toString();

    default:
      return null
    }
  };

  static extract_default_value(node) {
    switch (node.type) {
    case "str":
      return node.children[0];

    case "int":
    case "float":
      return node.children[0];

    case "true":
      return true;

    case "false":
      return false;

    case "nil":
      return null;

    default:
      return null
    }
  };

  // SQL generation methods
  static create_table_sql(stmt) {
    let columns_sql = stmt.columns.map(col => this.column_def_sql(col));

    // Add foreign key constraints
    for (let fk of stmt.foreign_keys.filter(x => x != null)) {
      columns_sql.push(`FOREIGN KEY (${fk.column}) REFERENCES ${fk.references_table}(${fk.references_column})`)
    };

    return `CREATE TABLE IF NOT EXISTS ${stmt.table} (\n  ${columns_sql.join(",\n  ")}\n);`
  };

  static column_def_sql(col) {
    let parts = [col.name];
    parts.push(this.sql_type(col.type));

    if (col.primaryKey) {
      parts.push("PRIMARY KEY");
      if (col.autoIncrement) parts.push("AUTOINCREMENT")
    };

    if ("null" in col && col.null == false) parts.push("NOT NULL");
    if ("default" in col) parts.push(`DEFAULT ${this.sql_value(col.default)}`);
    return parts.join(" ")
  };

  static sql_type(type) {
    switch (type.toString()) {
    case "string":
      return "TEXT";

    case "text":
      return "TEXT";

    case "integer":
      return "INTEGER";

    case "float":
    case "decimal":
      return "REAL";

    case "boolean":
      return "INTEGER";

    case "datetime":
    case "timestamp":
      return "TEXT";

    case "date":
      return "TEXT";

    case "time":
      return "TEXT";

    case "binary":
    case "blob":
      return "BLOB";

    default:
      return "TEXT"
    }
  };

  static sql_value(value) {
    switch (value) {
    case String:
      return `'${value.replaceAll("'", "''")}'`;

    case Integer:
    case Float:
      return value.toString();

    case true:
      return "1";

    case false:
      return "0";

    case null:
      return "NULL";

    default:
      return `'${value}'`
    }
  };

  static add_index_sql(stmt) {
    let unique = stmt.options.unique ? "UNIQUE " : "";
    let name = stmt.options.name ?? `index_${stmt.table}_on_${stmt.columns.join("_")}`;
    let columns = stmt.columns.join(", ");
    return `CREATE ${unique}INDEX IF NOT EXISTS ${name} ON ${stmt.table} (${columns});`
  };

  static add_column_sql(stmt) {
    return `ALTER TABLE ${stmt.table} ADD COLUMN ${stmt.column} ${this.sql_type(stmt.column_type)};`
  };

  static remove_column_sql(stmt) {
    return `-- Note: SQLite may not support DROP COLUMN\n-- ALTER TABLE ${stmt.table} DROP COLUMN ${stmt.column};`
  };

  static drop_table_sql(stmt) {
    return `DROP TABLE IF EXISTS ${stmt.table};`
  }
}
